source/common/ssl/context_config_impl.cc

```diff
@@ -16,20 +18,42 @@ namespace Ssl {
 
 namespace {
 
+std::string readSdsSecretName(const envoy::api::v2::auth::CommonTlsContext& config) {
+  return (!config.tls_certificate_sds_secret_configs().empty())
+             ? config.tls_certificate_sds_secret_configs()[0].name()
+             : EMPTY_STRING;
+}
+
+std::string readConfigSourceHash(const envoy::api::v2::auth::CommonTlsContext& config,
+                                 Secret::SecretManager& secret_manager) {
+  return (!config.tls_certificate_sds_secret_configs().empty() &&
+          config.tls_certificate_sds_secret_configs()[0].has_sds_config())
+             ? secret_manager.addOrUpdateSdsService(
+                   config.tls_certificate_sds_secret_configs()[0].sds_config(),
+                   config.tls_certificate_sds_secret_configs()[0].name())
+             : EMPTY_STRING;
+}
+
 std::string readConfig(
     const envoy::api::v2::auth::CommonTlsContext& config, Secret::SecretManager& secret_manager,
+    const std::string& config_source_hash, const std::string& secret_name,
     const std::function<std::string(const envoy::api::v2::auth::TlsCertificate& tls_certificate)>&
         read_inline_config,
-    const std::function<std::string(const Ssl::TlsCertificateConfig& secret)>& read_secret) {
+    const std::function<std::string(const Ssl::TlsCertificateConfig* secret)>&
```

Comment:
why did you changed this to pointer?

Reply:
Because SecretManagerImpl::findTlsCertificate() returns TlsCertificateConfig*, read_managed_secret(secret) takes a pointer instead of a unique_ptr. Do you want me to change it to TlsCertificateConfigPtr?

Reply:
No, you can still use `*` to deref the pointer, why I am asking this is this function cannot take a nullable, and you check if the pointer is null outside of this function.

Reply:
I change it back to read_managed_secret(*secret);, and the function takes const Ssl::TlsCertificateConfig& secret